pkgs/tools/backup/duplicity/default.nix

```diff
@@ -80,51 +82,50 @@ python3.pkgs.buildPythonApplication rec {
     par2cmdline # Add 'par2' to PATH.
   ] ++ lib.optionals stdenv.isLinux [
     util-linux # Add 'setsid' to PATH.
+  ] ++ lib.optionals stdenv.isDarwin [
+    getconf
   ] ++ (with python3.pkgs; [
     lockfile
     mock
     pexpect
     pytest
     pytest-runner
+    fasteners
   ]);
 
-  postInstall = ''
+  postInstall = let
+    binPath = lib.makeBinPath ([
+      gnupg
+      ncftp
+      rsync
+    ] ++ lib.optionals stdenv.isDarwin [
+      getconf
+    ]); in ''
     wrapProgram $out/bin/duplicity \
-      --prefix PATH : "${lib.makeBinPath [ gnupg ncftp rsync ]}"
+      --prefix PATH : "${binPath}"
   '';
 
   preCheck = ''
-    wrapPythonProgramsIn "$PWD/testing/overrides/bin" "$pythonPath"
-
-    # Add 'duplicity' to PATH for tests.
-    # Normally, 'setup.py test' adds 'build/scripts-2.7/' to PATH before running
-    # tests. However, 'build/scripts-2.7/duplicity' is not wrapped, so its
-    # shebang is incorrect and it fails to run inside Nix' sandbox.
-    # In combination with use-installed-scripts-in-test.patch, make 'setup.py
-    # test' use the installed 'duplicity' instead.
-    PATH="$out/bin:$PATH"
+    # tests need writable $HOME
+    HOME=$PWD/.home
 
-    # Don't run developer-only checks (pep8, etc.).
-    export RUN_CODE_TESTS=0
-
-    # check version string
-    duplicity --version | grep ${version}
-  '' + lib.optionalString stdenv.isDarwin ''
-    # Work around the following error when running tests:
-    # > Max open files of 256 is too low, should be >= 1024.
-    # > Use 'ulimit -n 1024' or higher to correct.
-    ulimit -n 1024
+    wrapPythonProgramsIn "$PWD/testing/overrides/bin" "$pythonPath"
   '';
 
-  # TODO: Fix test failures on macOS 10.13:
-  #
-  # > OSError: out of pty devices
-  doCheck = !stdenv.isDarwin;
+  doCheck = true;
+
+  passthru = {
+    tests.version = testers.testVersion {
+      package = self;
```

Comment:
Doesn't this (`package = self`) not work with overrides? I believe this scenario is what the `finalAttrs` pattern was made for.

(If this commit only did the version bump I could have merged now, but this extra change (`let self = ...; in self`) related to the version test makes me hesitate. Suggestion: keep each commit on topic.)

Reply:
I added this because I removed `duplicity --version | grep ${version}` from `preCheck.

I think you're right that it'll be broken w.r.t. overrides, but `buildPythonApplication` doesn't support `finalAttrs`.

Existing packages tend to do something like:

```
{ lib, python3, fetchPypi, khard, testers }:
[...]
  passthru.tests.version = testers.testVersion { package = khard; };
```

Which would also be unreliable in some cases, but it's at least a popular pattern.

Reply:
Actually, on further investigation, there are some places that do something similar to this.  One recent one:

https://github.com/ck3d/nixpkgs/blob/fe7b1a0628d4c20fdf54c6f0f1fd97488dd2f63f/pkgs/by-name/lo/local-ai/package.nix#L276

That would presumably because `buildGoModule` has the same limitation.

It should actually work for `override`, but not `overrideAttrs`.

Reply:
If you keep "duplicity --version | grep ${version}" in preCheck, it'll work either way (no matter how/what gets overridden).

Reply:
That test can't be done in the same way since `bin/duplicity` has been removed upstream.  It would have to do something like `duplicity/__main__.py --version`, or be moved to `installCheckPhase`.

I'm not a fan of those options because I think the real value in the test is verifying that the application works from outside the build environment.

I'm not opposed to removing it, but I would like to add some more tests in the future, at which point I'd have to deal with the test package reference anyway.